akka-projection-jdbc/src/main/scala/akka/projection/jdbc/internal/JdbcOffsetStore.scala

```diff
@@ -107,27 +109,68 @@ private[akka] class JdbcOffsetStore[S <: JdbcSession](
 
-      val now = new Timestamp(Instant.now(clock).toEpochMilli)
+    val storageReps = toStorageRepresentation(projectionId, offset)
 
-      def newRow(singleOffset: SingleOffset) = {
-        stmt.setString(InsertIndices.PROJECTION_NAME, singleOffset.id.name)
-        stmt.setString(InsertIndices.PROJECTION_KEY, singleOffset.id.key)
-        stmt.setString(InsertIndices.OFFSET, singleOffset.offsetStr)
-        stmt.setString(InsertIndices.MANIFEST, singleOffset.manifest)
-        stmt.setBoolean(InsertIndices.MERGEABLE, singleOffset.mergeable)
-        stmt.setTimestamp(InsertIndices.LAST_UPDATED, now)
-        stmt.addBatch()
-      }
+    val tryUpdates =
+      tryWithResource(conn.prepareStatement(settings.dialect.updateStatement())) { stmt =>
+
+        def updateRow(singleOffset: SingleOffset): Unit = {
+          // SET
+          stmt.setString(UpdateIndices.OFFSET, singleOffset.offsetStr)
+          stmt.setString(UpdateIndices.MANIFEST, singleOffset.manifest)
+          stmt.setBoolean(UpdateIndices.MERGEABLE, singleOffset.mergeable)
+          stmt.setTimestamp(UpdateIndices.LAST_UPDATED, now)
+          // WHERE
+          stmt.setString(UpdateIndices.PROJECTION_NAME, singleOffset.id.name)
+          stmt.setString(UpdateIndices.PROJECTION_KEY, singleOffset.id.key)
+
+          stmt.addBatch()
+        }
+
+        storageReps match {
+          case single: SingleOffset  => updateRow(single)
+          case MultipleOffsets(many) => many.foreach(updateRow)
+        }
+
+        stmt.executeBatch()
 
-      toStorageRepresentation(projectionId, offset) match {
-        case single: SingleOffset  => newRow(single)
-        case MultipleOffsets(many) => many.foreach(newRow)
       }
 
-      stmt.executeBatch()
+    // if something didn't get updated, insert it!!
+    if (tryUpdates.contains(0)) {
+      tryWithResource(conn.prepareStatement(settings.dialect.insertStatement())) { stmt =>
+
+        def insertRow(singleOffset: SingleOffset): Unit = {
 
+          // VALUES
+          stmt.setString(InsertIndices.PROJECTION_NAME, singleOffset.id.name)
+          stmt.setString(InsertIndices.PROJECTION_KEY, singleOffset.id.key)
+          stmt.setString(InsertIndices.OFFSET, singleOffset.offsetStr)
+          stmt.setString(InsertIndices.MANIFEST, singleOffset.manifest)
+          stmt.setBoolean(InsertIndices.MERGEABLE, singleOffset.mergeable)
+          stmt.setTimestamp(InsertIndices.LAST_UPDATED, now)
+
+          stmt.addBatch()
+        }
+
+        storageReps match {
+          case single: SingleOffset => insertRow(single)
+          case MultipleOffsets(many) =>
+            many.zipWithIndex.foreach {
+              case (offset, index) =>
+                if (tryUpdates(index) == 0) insertRow(offset)
+            }
+        }
+
+        // FIXME: we may want to let it crash if something is not inserted
+        // but this only make sense for exactly-once projections
```

Comment:
yes, but also for other projections it's good to see that it fails, and then something is wrong and we should restart

Reply:
My thought about this is that if we are using at-least-once and something happens, it that really worth to restart it. 

To be honest, I have no idea which kind of issue could come out of this. I would expect that if it's an error, we will have an exception. 

Anyway, let's just stop it and let it be reprocessed.